timeseries/src/autogluon/timeseries/metrics/abstract.py

```diff
@@ -0,0 +1,155 @@
+
+    def __str__(self) -> str:
+        return self.name
+
+    @property
+    def name_with_sign(self) -> str:
+        if self.greater_is_better:
+            prefix = ""
+        else:
+            prefix = "-"
+        return f"{prefix}{self.name}"
+
+    def __call__(
+        self,
+        data: TimeSeriesDataFrame,
+        predictions: TimeSeriesDataFrame,
+        prediction_length: int = 1,
+        target: str = "target",
+        seasonal_period: Optional[int] = None,
+        quantile_levels: Optional[List[float]] = None,
+        **kwargs,
+    ) -> float:
+        seasonal_period = get_seasonality(data.freq) if seasonal_period is None else seasonal_period
+
+        data_past = data.slice_by_timestep(None, -prediction_length)
+        data_future = data.slice_by_timestep(-prediction_length, None)
+
+        assert (predictions.num_timesteps_per_item() == prediction_length).all()
+        assert data_future.index.equals(predictions.index), "Prediction and data indices do not match."
+
+        self.save_past_metrics(
+            data_past=data_past,
+            target=target,
+            seasonal_period=seasonal_period,
+            **kwargs,
+        )
+        with warning_filter():
+            metric_value = self.compute_metric(
+                data_future=data_future,
+                predictions=predictions,
+                target=target,
+                quantile_levels=quantile_levels,
+                **kwargs,
+            )
+        self.clear_past_metrics()
+        return metric_value * self.sign
+
+    score = __call__
+
+    def compute_metric(
+        self,
+        data_future: TimeSeriesDataFrame,
+        predictions: TimeSeriesDataFrame,
+        target: str = "target",
+        quantile_levels: Optional[List[float]] = None,
+        **kwargs,
+    ) -> float:
+        """Internal method that computes the metric for given forecast & actual data.
+
+        This method should be implemented by all custom metrics.
+
+        Parameters
+        ----------
+        data_future : TimeSeriesDataFrame
+            Actual values of the time series during the forecast horizon (``prediction_length`` values for each time
+            series in the dataset). This data frame is guaranteed to have the same index as ``predictions``.
+        predictions : TimeSeriesDataFrame
+            Data frame with predictions for the forecast horizon. Contain columns "mean" (point forecast) and the
+            columns corresponding to each of the quantile levels.
+        target : str, default = "target"
+            Name of the column in ``data_future`` that contains the target time series.
+
+        Returns
+        -------
+        score : float
+            Value of the metric for given forecast and data. If self.greater_is_better is True, returns score in
+            greater-is-better format, otherwise in lower-is-better format.
+
+        """
+        raise NotImplementedError
+
+    def save_past_metrics(
```

Comment:
The reason this method exists is because it's used in the Ensemble: https://github.com/autogluon/autogluon/blob/master/timeseries/src/autogluon/timeseries/models/ensemble/greedy_ensemble.py#L64 to speed up repeated metric evaluations.

We could make this method "private" in a sense that it's "private to AutoGluon" (other objects use it throughout the library); but it wouldn't be "private like in Java" (where only the TimeSeriesScorer object uses this method internally), as that would break the ensemble functionality.

Do you think that designating the method "private to AutoGluon" would justify renaming it to `_save_past_metrics`?

Reply:
Hmm. I see, then it isn't private.